all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,526 @@
+
+Alternatively, you can download it as a configuration file:
+
+    <some button or something that downloads a file name temporalio-client.toml>
+
+You can place this configuration file in the default place used by CLIs/SDKs:
+
+* macOS - `$HOME/Library/Application Support/temporalio/temporal-client.toml`
+* Windows - `%AppData%/temporalio/temporal-client.toml` (often `C:\Users\<myuser>\AppData\Roaming\temporalio\temporal-client.toml`)
+* Linux - `$HOME/.config/temporalio/temporal-client.toml`
+
+This will be automatically consumed by the SDKs/CLI. This file can also be put anywhere else and `TEMPORAL_CONFIG_FILE`
+environment variable can be set to point to it.
+```
+
+#### Possible CLI Creation Flow
+
+Assuming cloud support is present in `temporal` CLI, you can have the normal certificate generation steps, e.g.
+
+```
+temporal cloud gen ca ...
+temporal cloud gen leaf ...
+temporal cloud namespace add-accepted-client-ca ...
+```
+
+And that last step can by default set a `temporal-cloud` profile in the configuration with the address, namespace, and
+mTLS certificates. So now all you have to do to use cloud in CLI is either `--profile temporal-cloud` or set
+`TEMPORAL_PROFILE=temporal-cloud` environment variable. Granted you may want better certificate management than that and
+whether this is done by default can be discussed, but the idea is the same.
+
+#### Multi-Namespace Config Use
+
+Some users may want the same configuration for multiple namespaces (or multiple addresses or multiple API keys or
+whatever). Users can supply partial configuration. For instance, you can have this config:
+
+```toml
+[profile.temporal-cloud]
+address = "my-dev-ns.a1b2c.tmprl.cloud:7233"
+tls.client_cert_path = "path/to/my/dev-cert.pem"
+tls.client_key_path = "path/to/my/dev-cert.pem"
+```
+
+And then still use CLI like
+
+    temporal workflow list --namespace my-specific-namespace
+
+This will use the config file information for all but the namespace. Similarly you could just set the
+`TEMPORAL_NAMESPACE` env var to `my-specific-namespace` and run:
+
+    temporal workflow list
+
+Similarly, if you're using an SDK, you can load the config and just switch namespaces in code. So you can have:
+
+```python
+config = envconfig.load_client_connect_config()
+client = await Client.connect(namespace="my-specific-namespace", **config)
+```
+
+## Specification
+
+### Values and File Format
+
+* The format is a TOML object with the following fields:
+  * `profile` - Object with keys as profile names. The default profile should be named `default`.
+    * `<name>` - Profile object with configuration values. Profile names are case-insensitive. 💭 Why? They are needed
+      for environment variables which are case insensitive. It is a validation error to have a config file with two
+      separately-cased profile names that are equal case-insensitively.
+      * `address` - Client address, aka gRPC "host:port". This cannot be a URL, it must be `host:port`.
```

Comment:
Let's be specific about IPv6 address, e.g.: `[ipv6address]:port`

Reply:
> Looking back, I will probably back off validation here and just say "address passed to the underlying client, we encourage host:port".

I disagree. `temporal config set` should validate for what is expected to be the correct format. Not doing so will make this a common source of errors for new users for a benefit that only matters for a very tiny percentage of "expert users", who could anyway still provide "out-of-norm" values by editing the config file by hand.

Reply:
So what do you believe the validation for `address` should be if I want to have Go allow a type of value that, say, Python doesn't support, such as a custom resolver? To validate this value properly means to assume how an SDK uses the value or assume that all of our SDKs share a common validation of this value which is not accurate.

Reply:
Validation in CLI should be based on a least-common-denominator format that we guarantee to be supported in every SDKs to support.

IMO, that should be:

```
ipv4:port
hotname:port
[ipv6]:port
```

Anything more than that should not be accepted as input on `temporal config set`. That way, we promote using things that will work in all SDKs.

However, I don't think we need the _SDK side_ to validate that. So if a user edit his config file and manually puts a value of another form, the SDK will just pass that to the grpc library, and that may work or that may not.